mtags/src/main/scala-3/scala/meta/internal/pc/completions/NamedArgCompletions.scala

```diff
@@ -216,4 +299,91 @@ object NamedArgCompletions:
     ) ::: findPossibleDefaults() ::: fillAllFields()
   end contribute
 
+  extension (method: Symbols.Symbol)
+    def vparamss(using Context) = method.filteredParamss(_.isTerm)
+    def tparams(using Context) = method.filteredParamss(_.isType).flatten
+    def filteredParamss(f: Symbols.Symbol => Boolean)(using Context) =
+      method.paramSymss.filter(params => params.forall(f))
 end NamedArgCompletions
+
+class FuzzyArgMatcher(tparams: List[Symbols.Symbol])(using Context):
+
+  /**
+   * A heuristic for checking if the passed arguments match the method's arguments' types.
+   * For non-polymorphic methods we use the subtype relation (`<:<`)
+   * and for polymorphic methods we use a heuristic.
+   * We check the args types not the result type.
+   */
+  def doMatch(
+      allArgsProvided: Boolean
+  )(expectedArgs: List[Symbols.Symbol], actualArgs: List[Tree]) =
+    (expectedArgs.length == actualArgs.length ||
+      (!allArgsProvided && expectedArgs.length >= actualArgs.length)) &&
+      actualArgs.zipWithIndex.forall {
+        case (i: Ident, _) => true
+        case (NamedArg(name, arg), _) =>
+          expectedArgs.exists { expected =>
+            expected.name == name && (!arg.hasType || arg.typeOpt.unfold
+              .fuzzyArg_<:<(expected.info))
+          }
+        case (arg, i) =>
+          !arg.hasType || arg.typeOpt.unfold.fuzzyArg_<:<(expectedArgs(i).info)
+      }
+
+  extension (arg: Type)
+    def fuzzyArg_<:<(expected: Type) =
+      if tparams.isEmpty then arg <:< expected
+      else
+        val substExpected = tryToMakeTypeParamsSubstitutions(expected, arg)
+        substExpected.exists(arg <:< _)
+    def unfold =
+      arg match
+        case arg: TermRef => arg.underlying
+        case e => e
+
+  private def tryToMakeTypeParamsSubstitutions(
+      t: Type,
+      arg: Type,
+  ): Option[Type] =
+    t match
+      case e if tparams.exists(_ == e.typeSymbol) =>
+        val matchingParam = tparams.find(_ == e.typeSymbol).get
+        matchingParam.info match
+          case TypeBounds(lo, hi) =>
+            if lo <:< arg && arg <:< hi then Some(arg)
+            else None
+          case _ => Some(arg)
+      case OrType(e1, e2)
+          if tparams.exists(t => t == e1.typeSymbol || t == e2.typeSymbol) =>
+        Some(arg)
+      case AndType(e1, e2) if tparams.exists(_ == e1.typeSymbol) => Some(e2)
+      case AndType(e1, e2) if tparams.exists(_ == e2.typeSymbol) => Some(e1)
+      case o @ OrType(e1, e2) =>
+        for
+          e1s <- tryToMakeTypeParamsSubstitutions(e1, arg)
+          e2s <- tryToMakeTypeParamsSubstitutions(e2, arg)
+        yield OrType(e1s, e2s, o.isSoft)
+      case AndType(e1, e2) =>
+        for
+          e1s <- tryToMakeTypeParamsSubstitutions(e1, arg)
+          e2s <- tryToMakeTypeParamsSubstitutions(e2, arg)
+        yield AndType(e1s, e2s)
+      case AppliedType(et, eparams) =>
```

Comment:
Can you add a comment explaining this part a bit?

Reply:
I discovered `WildcardType` (I knew smth of the sort had to exist 😛) and simplified it massively.

Reply:
This looks way better!